### PR TITLE
Extend server configuration with compression config

### PR DESCRIPTION
**Pull request by numinnex**

### CODE REVIEW

File: server/src/configs/compression_algorithm.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use serde_json::json;
+
+    #[test]
+    fn test_serialize() {
+        let producer_alg = CompressionAlgorithm::Producer;
+        let producer_serialized = serde_json::to_string(&producer_alg).unwrap();
```
Comment (spetz) [inline server/src/configs/compression_algorithm.rs]: Should we use `serde_json`, as it seems that it's a basic mapping between an enum and string types? Maybe just the usage of `Into` and `From` traits would suffice? Unless it's supposed to be extended in the future and contain additional fields (e.g. as a struct).

File: server/src/configs/compression_algorithm.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        let gzip_alg = CompressionAlgorithm::Gzip;
+        let gzip_serialized = serde_json::to_string(&gzip_alg).unwrap();
+        assert_eq!(producer_serialized, json!("producer").to_string());
+        assert_eq!(gzip_serialized, json!("gzip").to_string());
+    }
+
+    #[test]
+    fn test_deserialize() {
+        let json_data = "\"producer\"";
+        let deserialized: Result<CompressionAlgorithm, serde_json::Error> =
```
Comment (spetz) [inline server/src/configs/compression_algorithm.rs]: Same story as with the serialization :)

File: iggy/src/compression.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+impl From<CompressionKind> for String {
+    fn from(value: CompressionKind) -> Self {
+        match value {
+            CompressionKind::None => "none".to_string(),
+            CompressionKind::Gzip => "gzip".to_string(),
+        }
+    }
+}
+struct CompressionKindVisitor;
```
Comment (spetz) [inline iggy/src/compression.rs]: Is it still needed along with the `Visitor`, given that the `serde` it not used anymore?

**Review by spetz — DISMISSED** [review]

File: iggy/src/compression.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(CompressionKindVisitor)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (hubcio) [inline iggy/src/compression.rs]: please add negative (cold path) subcases

**Review by hubcio — COMMENTED** [review]
Please also add warning print if compression is enabled, something like:
Server started with compression enabled, but this feature is not yet implemented!

you can do it in validate() for ServerConfig

File: iggy/src/compression.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    None,
+    Gzip,
+}
+
+impl FromStr for CompressionKind {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "Gzip" | "gzip" => Ok(CompressionKind::Gzip),
```
Comment (hubcio) [inline iggy/src/compression.rs]: please use https://doc.rust-lang.org/std/primitive.str.html#method.to_lowercase

File: server/src/configs/validators.rs [hunk 6]
```diff
@@ -14,12 +16,28 @@ impl Validatable<ServerError> for ServerConfig {
...
+        self.system.compression.validate()?;
         self.personal_access_token.validate()?;
 
         Ok(())
     }
 }
 
+impl Validatable<ServerError> for CompressionConfig {
+    fn validate(&self) -> Result<(), ServerError> {
+        let compression_alg = self.default_algorithm.clone();
```
Comment (hubcio) [inline server/src/configs/validators.rs]: use reference, don't clone

File: iggy/src/compression.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+use crate::error::Error;
+
+// for now only those, in the future will add snappy, lz4, zstd (same as in confluent kafka) in addition to that
+// we should consider brotli as well.
+#[derive(Debug, PartialEq, Clone)]
+pub enum CompressionKind {
```
Comment (hubcio) [inline iggy/src/compression.rs]: change to CompressionAlgorithm

**Review by hubcio — APPROVED** [review]

**Pull request merged**